File: problem/Min_Powered_City/maxPower.py
class Solution(object):
    def maxPower(self, stations, r, k):
        prefix = [0]
        size = len(stations)
        for i in stations:
            prefix.append(prefix[-1] + i)
        for i in range(size):
            stations[i] = prefix[min(i+r+1, size)] - prefix[max(0, i-r)]

        def check(min_power):
            diff = [0] * size
            cur_diff = 0
            cnt_station = 0
            for i, power in enumerate(stations):
                cur_diff += diff[i]
                power_diff = min_power - power - cur_diff

                if power_diff > 0: 
                    cnt_station += power_diff
                    if cnt_station > k:
                        return False
                    cur_diff += power_diff
                    if i + 2 * r + 1 < size: diff[i + 2 * r + 1] -= power_diff

            return True
        left = min(stations)
        right = left + k
        while left <= right:
            mid = left + (right - left) // 2
            if check(mid): left = mid + 1
            else: right = mid - 1

        return right

File: problem/Min_Powered_City/maxPower.c
static bool canAchieve(long long x, int *stations, long long *power, int n, int r, long long k)
{
    long long *diff = (long long *)calloc(n + 1, sizeof(long long));
    long long added = 0;
    for (int i = 0; i < n; i++)
    {
        added += diff[i];
        long long curr = power[i] + added;
        if (curr < x)
        {
            long long need = x - curr;
            if (need > k)
            {
                free(diff);
                return false;
            }
            k -= need;
            added += need;
            int end = i + 2 * r + 1;
            if (end < n)
                diff[end] -= need;
        }
    }
    free(diff);
    return true;
}
long long maxPower(int *stations, int n, int r, int k)
{
    long long *prefix = (long long *)calloc(n + 1, sizeof(long long));
    long long *power = (long long *)malloc(sizeof(long long) * n);
    for (int i = 0; i < n; i++)
        prefix[i + 1] = prefix[i] + stations[i];
    for (int i = 0; i < n; i++)
    {
        int left = (i - r < 0) ? 0 : i - r;
        int right = (i + r >= n) ? n - 1 : i + r;
        power[i] = prefix[right + 1] - prefix[left];
    }
    long long lo = 0, hi = prefix[n] + k, ans = 0;
    while (lo <= hi)
    {
        long long mid = lo + (hi - lo) / 2;
        if (canAchieve(mid, stations, power, n, r, k))
        {
            ans = mid;
            lo = mid + 1;
        }
        else
        {
            hi = mid - 1;
        }
    }
    free(prefix);
    free(power);
    return ans;
}